src/camera.py

```diff
@@ -0,0 +1,152 @@
+        return None
+
+    def _url_is_valid(self) -> bool:
+        """Checks if the stream URL is valid."""
+        if self.stream_url is None:
+            return False
+
+        return all(
+            [
+                self.stream_url != "/static/no.jpg",
+                "?stream" not in self.stream_url,
+                ".jpg" in self.stream_url,
+            ]
+        )
+
+    def _get_camera_details(self) -> Optional[str]:
+        """Extracts the camera details."""
+        details = self.page_tree.xpath('//div[@class="camera-details"]')
+        details_tuple = tuple(detail.text_content() for detail in details)
+        details = "".join(detail.replace("\n", "").replace("\t", "").strip() for detail in details_tuple)
+        return details
+
+    def _parse_camera_details(self) -> Optional[Dict[str, str]]:
+        """Parses the camera details and returns the camera info as a dictionary."""
+
+        if not self.details:
+            return None
+
+        details = self.details
+
+        city_start = details.find("City: ") + len("City: ")
+        city_end = details.find("Latitude:")
+        city = details[city_start:city_end]
+
+        region_start = details.find("Region:") + len("Region:")
+        region_end = details.find("City:")
+        region = details[region_start:region_end]
+
+        country_start = details.find("Country:") + len("Country:")
+        country_end = details.find("Country code:")
+        country = details[country_start:country_end]
+
+        country_code_start = details.find("Country code:") + len("Country code:")
+        country_code_end = details.find("Region:")
+        country_code = details[country_code_start:country_code_end]
+
+        return {
+            "city": city,
+            "region": region,
+            "country": country,
+            "country_code": country_code,
+        }
+
+    @exponential_backoff(
+        attempts=s.CAMERA_FETCH_ATTEMPTS,
+        initial_delay=s.INITIAL_RETRY_DELAY,
+        factor=s.RETRY_DELAY_FACTOR,
+        exception_types=(SaveImageException, FetchCamerasException),
+    )
+    def _save_image(self, image_file_path: str, camera_url: str, request_headers: Dict[str, str]) -> bool:
+        """Saves the image from the camera stream URL."""
+        try:
+            r = requests.get(camera_url, headers=request_headers, timeout=s.IMAGE_SAVE_TIMEOUT)
+            r.raise_for_status()
+
+            with open(image_file_path, "wb") as f:
+                f.write(r.content)
+            return True
+        except (RequestException, ReadTimeout) as e:
+            logger.error(f"error saving image: {e}")
+            raise SaveImageException(f"error saving image: {e}")
+
+    def _image_is_solid_color(self, image_file_path: str) -> bool:
+        """Checks if the image consists of a single color."""
+        image = cv2.imread(image_file_path)
+
+        if image is None:
+            logging.error("image is empty. skipping...")

```

Comment:
You probably want the file path in the log here if you have access to it so you know in your logs which was empty.